californium-core/src/main/java/org/eclipse/californium/core/network/stack/BlockwiseLayer.java

```diff
@@ -248,6 +248,7 @@ private Request startBlockwiseUpload(final Exchange exchange, final Request requ
 			status = getOutboundBlock1Status(key, exchange, request);
 
 			final Request block = status.getNextRequestBlock();
+			block.setResponseTimeout(blockTimeout);
```

Comment:
If that's a "single block request", why should that large timeout be used?

Reply:
What do you mean by large timeout ? you talk about the parent request timeout ? or you just found that `blocktTimeout` is a too large value ?

Reply:
FMPOV the intention of the `blockTimeout` is different and therefore the value seems for me to be rather high.

Reply:
The default value for blockTimeout is 30s
An exchange with piggypack response is MAX_TRANSMIT_WAIT => 97s.
The [spec](https://tools.ietf.org/html/rfc7959#section-2.4) say : 
_"Clients that want to retrieve all blocks of a resource SHOULD strive to do so without undue delay. Servers can fully expect to be free to discard any cached state after a period of EXCHANGE_LIFETIME ([RFC7252], Section 4.8.2) after the last access to the state, however, there is no requirement to always keep the state for as long."_
(same thing for [block1](https://tools.ietf.org/html/rfc7959#section-2.5))
EXCHANGE_LIFETIME is about 247s

Before this PR, `blockTimeout` was about server which does not receive request in time.
Using it  for a client which does not receive response seems not so bad.
We could imagine to add a new configuration entry, but I found this a bit overkill and add complexity for users (maybe the price of flexibility...)

Reply:
> The default value for blockTimeout is 30s

I see also a value of 5 minutes in NetworkConfigDefaults.

Reply:
I looked at the BlockwiseLayer class ... 
![It's a trap for developer :p](https://images.duckduckgo.com/iu/?u=https%3A%2F%2Fmedia.tenor.co%2Fimages%2F5f453103d56669d5edc16f3858997478%2Fraw&f=1)
I suppose we should set the same value for both (ideally only one constant)

We should choose between 30s and 5min (or something else).

No matter which value we choose, does it make sense to you to use the same value for missing request and missing response for transparent block transfer ?

Reply:
FMPOV:
I would stick with the value of the resonseTimeout from the Request as timeout for single blocks (last pending request).

Reply:
And with the 5 minutes for the state cleanup, but not for the response timeout.

Reply:
> I would stick with the value of the resonseTimeout from the Request as timeout for single blocks (last pending request).

I didn't understand.

Could you argue why you consider time between request and response should be different than time between response and request ?

Reply:
FMPOV, the blocktime out is for more a caching time on the server side. The server can't trigger the client to send the next block request. Their is no specification saying, that the next block request must be within a time.

For the reponse timeout this seems to be slightly different. Yes with a separate response there will also be no specified timeout, but as user of the API I would prefer, that the value I had provided is used.

Reply:
> Their is no specification saying, that the next block request must be within a time.

As the spec not saying, that a response must be within a time.

>  I would prefer, that the value I had provided is used.

The API could say that the value for block response timeout is  BLOCKWISE_STATUS_LIFETIME. So the timeout will be value you provide.

But ok let's explore the other solution. Just focus on transparent block transfer.
If I well understand you prefer to always use the "parent" request timeout for each block ?
E.g. I send a GET request with response timeout set to 5minutes, each block will have a 5minutes response timeout ? that's what you mean ?

Reply:
Do you want to used the "blocktime", because you usually don't want to provide the response timeout?

Reply:
I ment, provide as response time at the request not as provided blocktime in the configration.

Reply:
> If I well understand you prefer to always use the "parent" request timeout for each block ?

yes.

> E.g. I send a GET request with response timeout set to 5minutes, each block will have a 5minutes response timeout ? that's what you mean ?

yes, but though it is a single block request, I would recommend a smaller value (but that's up to the caller).

Reply:
> Do you want to used the "blocktime", because you usually don't want to provide the response timeout?

I wanted to let the possibility to not set timeout to keep a kind of backward compatibility, but this is not mandatory.

Ok let's continue with this example.
User send a GET request and set its timeout to 5minutes. 
In my mind that's mean he didn't want to wait more than 5 minutes for his response.
User can not know if transparent block mode will be used. (This is up to the server)
If for each block request we reuse the same timer than the original/parent GET. This timer will never be used as the original/parent timer will alway be reached first.

In that case why using response timeout for block... but as I explained above for block notification there is no "parent/original" request. So we need to set a response timeout for each block.
So we can do a special case for this one. What could it be ?
- Using the timeout set on the observe request for each block ? maybe a bit strange. We should choose a default value if there is no timeout on this observe request.
- or, using BLOCKWISE_STATUS_LIFETIME just for notification.
- or new config parameters.

Not sure all of this will be clearer for users.
FMPOV,  this is clearer something like "all transparent block exchange should take at least LOCKWISE_STATUS_LIFETIME".

Reply:
>  I wanted to let the possibility to not set timeout to keep a kind of backward compatibility, but this is not mandatory.

My question was imprecise: though the blocktime is always used as response time, even if 0 was used for the request, I thought that's your intention.

Reply:
In the current implementation, for block request we always use BLOCKWISE_STATUS_LIFETIME even if parent/original request as a response timeout set to 0.

Reply:
Assuming unintended transparent blockwise transfer together with a response time out will fail.

Either you aware of that and provide a proper value, or you will be faced too many response timeouts.

If we want a "reliable" function, then the response timeout must be the timeout of a single response, and yes , maybe changing the original expectation. But that's an issue of this transparent blockwise mode at all.

And I would go for only one scheduled timeout task and refresh that on every block, as the blockwise cleanup does also.

Reply:
> in that case why using response timeout for block... but as I explained above for block notification there is no "parent/original" request. So we need to set a response timeout for each block.

When we create the requests for the followup transfer, we have the clone of the original request. We may require to serialize/persist the response timeout as well.

Reply:
Sorry, I still get the strong feeling that using the blocktime is intended as always response timeout for blockwise transfer. 
If so, (because using the response time of the request may be not that easy,) I think using the 1. approach in your comment

https://github.com/eclipse/californium/pull/418#issuecomment-335461350

launch a cleaning task for blockwise exchanges would be better.

Reply:
> In the current implementation, for transparent block request we always use BLOCKWISE_STATUS_LIFETIME even if parent/original request as a response timeout set to 0.

I read your comment too late :-). But with that, I think the cleaner task will do a better job.

Reply:
why ... ? it just do the same thing without the benefits of adding an optional timeout response for all request... 

By the way, I ask before which solution you prefer and you say you have no preference... now you say you prefer the other one without any clear arguments (I mean if there is arguments I can understand but I don't see any) ... 

Sry I loose a bit my temper, but I feel this is more and more difficult to work on this project...

Reply:
My answer in

https://github.com/eclipse/californium/pull/418#issuecomment-335710910

includes "Please extend such a timeout on every block, not that a large latency accidentally cleansup an ongoing blockwise exchange."

Reply:
I do not understand go for 1. solution.